In [2]:
import pandas as pd
import numpy as np

###### I've tried to download the file and I've found it is too heavy to download it directly, so lo lets download first 10 rows

In [3]:
bank = pd.read_csv('./data/train_ver2.csv', nrows=10)
bank.head()

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  fecha_alta  \
0  2015-01-28   1375586            N              ES    H   35  2015-01-12   
1  2015-01-28   1050611            N              ES    V   23  2012-08-10   
2  2015-01-28   1050612            N              ES    V   23  2012-08-10   
3  2015-01-28   1050613            N              ES    H   22  2012-08-10   
4  2015-01-28   1050614            N              ES    V   23  2012-08-10   

   ind_nuevo  antiguedad  indrel  ...  ind_hip_fin_ult1  ind_plan_fin_ult1  \
0          0           6       1  ...                 0                  0   
1          0          35       1  ...                 0                  0   
2          0          35       1  ...                 0                  0   
3          0          35       1  ...                 0                  0   
4          0          35       1  ...                 0                  0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1  ind_valo_fin_ult1  \
0                 0                 0                 0                  0   
1                 0                 0                 0                  0   
2                 0                 0                 0                  0   
3                 0                 0                 0                  0   
4                 0                 0                 0                  0   

  ind_viv_fin_ult1 ind_nomina_ult1  ind_nom_pens_ult1  ind_recibo_ult1  
0                0               0                  0                0  
1                0               0                  0                0  
2                0               0                  0                0  
3                0               0                  0                0  
4                0               0                  0                0  

[5 rows x 48 columns]

In [4]:
bank.columns

Index(['fecha_dato', 'ncodpers', 'ind_empleado', 'pais_residencia', 'sexo',
       'age', 'fecha_alta', 'ind_nuevo', 'antiguedad', 'indrel',
       'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'indresi', 'indext',
       'conyuemp', 'canal_entrada', 'indfall', 'tipodom', 'cod_prov',
       'nomprov', 'ind_actividad_cliente', 'renta', 'segmento',
       'ind_ahor_fin_ult1', 'ind_aval_fin_ult1', 'ind_cco_fin_ult1',
       'ind_cder_fin_ult1', 'ind_cno_fin_ult1', 'ind_ctju_fin_ult1',
       'ind_ctma_fin_ult1', 'ind_ctop_fin_ult1', 'ind_ctpp_fin_ult1',
       'ind_deco_fin_ult1', 'ind_deme_fin_ult1', 'ind_dela_fin_ult1',
       'ind_ecue_fin_ult1', 'ind_fond_fin_ult1', 'ind_hip_fin_ult1',
       'ind_plan_fin_ult1', 'ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1'],
      dtype='object')

In [5]:
# I've used this to get the dataframe by column and, with the three cells below, determine the optimal cell type.
# The purpose of doing that is open all dataframe for feature enginering.

usecols = ['ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1']

bank_by_cols = pd.read_csv('./data/train_ver2.csv', usecols=usecols
                         , na_values={'na':0, 'NA':0 , 'NaN':0.0, 'nan':0.0}
                        )


In [6]:
# This helps to reduce the bytes of integers to minimum (int8) if possible

for int_col in bank_by_cols.select_dtypes('integer').columns:
    bank_by_cols[int_col] = pd.to_numeric(bank_by_cols[int_col], downcast='integer')
    
bank_by_cols.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13647309 entries, 0 to 13647308
Data columns (total 8 columns):
ind_pres_fin_ult1    int8
ind_reca_fin_ult1    int8
ind_tjcr_fin_ult1    int8
ind_valo_fin_ult1    int8
ind_viv_fin_ult1     int8
ind_nomina_ult1      float64
ind_nom_pens_ult1    float64
ind_recibo_ult1      int8
dtypes: float64(2), int8(6)
memory usage: 286.3 MB


In [7]:
# same for float

for int_col in bank_by_cols.select_dtypes('float').columns:
    bank_by_cols[int_col] = pd.to_numeric(bank_by_cols[int_col], downcast='float')
    
bank_by_cols.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13647309 entries, 0 to 13647308
Data columns (total 8 columns):
ind_pres_fin_ult1    int8
ind_reca_fin_ult1    int8
ind_tjcr_fin_ult1    int8
ind_valo_fin_ult1    int8
ind_viv_fin_ult1     int8
ind_nomina_ult1      float32
ind_nom_pens_ult1    float32
ind_recibo_ult1      int8
dtypes: float32(2), int8(6)
memory usage: 182.2 MB


In [8]:
# Convert object to category - Will be used later

for cat_feature in bank_by_cols.select_dtypes('object').columns:
    bank_by_cols[cat_feature] = bank_by_cols[cat_feature].astype('category')

bank_by_cols.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13647309 entries, 0 to 13647308
Data columns (total 8 columns):
ind_pres_fin_ult1    int8
ind_reca_fin_ult1    int8
ind_tjcr_fin_ult1    int8
ind_valo_fin_ult1    int8
ind_viv_fin_ult1     int8
ind_nomina_ult1      float32
ind_nom_pens_ult1    float32
ind_recibo_ult1      int8
dtypes: float32(2), int8(6)
memory usage: 182.2 MB


In [9]:
# Dictionary with all columns and its types, this is made in order to download the full file

cols_types = {'fecha_dato':'object',
'ncodpers':'int32',
'ind_empleado':'object',
'pais_residencia':'object',
'sexo':'object',
'age':'object',          # este tiene un valor NA, en la doc de pandas se puede decir NA=0
'fecha_alta':'object',
'ind_nuevo':'float32',
'antiguedad':'object',   # este tiene un valor NA, en la doc de pandas se puede decir NA=0
'indrel':'float32',
              
'ult_fec_cli_1t':'object',
'indrel_1mes':'object',
'tiprel_1mes':'object',
'indresi':'object',
'indext':'object',
'conyuemp':'object',
'canal_entrada':'object',
'indfall':'object',
'tipodom':'float32',
'cod_prov':'float32',    # debería ser int32
              
'nomprov':'object',
'ind_actividad_cliente':'float32',
'renta':'float32',
'segmento':'object',
'ind_ahor_fin_ult1':'int8',
'ind_aval_fin_ult1':'int8',
'ind_cco_fin_ult1':'int8',
'ind_cder_fin_ult1':'int8',
'ind_cno_fin_ult1':'int8',
'ind_ctju_fin_ult1':'int8',
              
'ind_ctma_fin_ult1':'int8',
'ind_ctop_fin_ult1':'int8',
'ind_ctpp_fin_ult1':'int8',
'ind_deco_fin_ult1':'int8',
'ind_deme_fin_ult1':'int8',
'ind_dela_fin_ult1':'int8',
'ind_ecue_fin_ult1':'int8',
'ind_fond_fin_ult1':'int8',
'ind_hip_fin_ult1':'int8',
'ind_plan_fin_ult1':'int8',
              
'ind_pres_fin_ult1':'int8',
'ind_reca_fin_ult1':'int8',
'ind_tjcr_fin_ult1':'int8',
'ind_valo_fin_ult1':'int8',
'ind_viv_fin_ult1':'int8',
'ind_nomina_ult1':'float32',
'ind_nom_pens_ult1':'float32',
'ind_recibo_ult1':'int8'
             }

In [10]:
bank_total = pd.read_csv('./data/train_ver2.csv', dtype=cols_types
                         , na_values={'na':0, 'NA':0 , 'NaN':0.0, 'nan':0.0}
                        )
bank_total.head()

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  fecha_alta  \
0  2015-01-28   1375586            N              ES    H   35  2015-01-12   
1  2015-01-28   1050611            N              ES    V   23  2012-08-10   
2  2015-01-28   1050612            N              ES    V   23  2012-08-10   
3  2015-01-28   1050613            N              ES    H   22  2012-08-10   
4  2015-01-28   1050614            N              ES    V   23  2012-08-10   

   ind_nuevo antiguedad  indrel  ... ind_hip_fin_ult1 ind_plan_fin_ult1  \
0        0.0          6     1.0  ...                0                 0   
1        0.0         35     1.0  ...                0                 0   
2        0.0         35     1.0  ...                0                 0   
3        0.0         35     1.0  ...                0                 0   
4        0.0         35     1.0  ...                0                 0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1 ind_valo_fin_ult1  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

  ind_viv_fin_ult1 ind_nomina_ult1  ind_nom_pens_ult1  ind_recibo_ult1  
0                0             0.0                0.0                0  
1                0             0.0                0.0                0  
2                0             0.0                0.0                0  
3                0             0.0                0.0                0  
4                0             0.0                0.0                0  

[5 rows x 48 columns]

<font color=blue>

### Finally I got open the file!!
</font>

Lets check the size of the dataframe


In [11]:
bank_total.shape

(13647309, 48)

Now I want to know which columns are categories.

For this purpose I've create a function that prints the number of unique values and the unique values. I prefer doing it on this way since I'm able to check all columns in one step. It consumes more computational resources but I think is worthly.

In [12]:
all_columns = ['fecha_dato', 'ncodpers', 'ind_empleado', 'pais_residencia', 'sexo',
       'age', 'fecha_alta', 'ind_nuevo', 'antiguedad', 'indrel',
       'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'indresi', 'indext',
       'conyuemp', 'canal_entrada', 'indfall', 'tipodom', 'cod_prov',
       'nomprov', 'ind_actividad_cliente', 'renta', 'segmento',
       'ind_ahor_fin_ult1', 'ind_aval_fin_ult1', 'ind_cco_fin_ult1',
       'ind_cder_fin_ult1', 'ind_cno_fin_ult1', 'ind_ctju_fin_ult1',
       'ind_ctma_fin_ult1', 'ind_ctop_fin_ult1', 'ind_ctpp_fin_ult1',
       'ind_deco_fin_ult1', 'ind_deme_fin_ult1', 'ind_dela_fin_ult1',
       'ind_ecue_fin_ult1', 'ind_fond_fin_ult1', 'ind_hip_fin_ult1',
       'ind_plan_fin_ult1', 'ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1']
for col in all_columns:
    print(col, '- num unique values:', bank_total[col].nunique(), '\n', bank_total[col].unique(), '\n\n')

fecha_dato - num unique values: 17 
 ['2015-01-28' '2015-02-28' '2015-03-28' '2015-04-28' '2015-05-28'
 '2015-06-28' '2015-07-28' '2015-08-28' '2015-09-28' '2015-10-28'
 '2015-11-28' '2015-12-28' '2016-01-28' '2016-02-28' '2016-03-28'
 '2016-04-28' '2016-05-28'] 


ncodpers - num unique values: 956645 
 [1375586 1050611 1050612 ... 1173729 1164094 1550586] 


ind_empleado - num unique values: 5 
 ['N' nan 'A' 'B' 'F' 'S'] 


pais_residencia - num unique values: 118 
 ['ES' nan 'CA' 'CH' 'CL' 'IE' 'AT' 'NL' 'FR' 'GB' 'DE' 'DO' 'BE' 'AR' 'VE'
 'US' 'MX' 'BR' 'IT' 'EC' 'PE' 'CO' 'HN' 'FI' 'SE' 'AL' 'PT' 'MZ' 'CN'
 'TW' 'PL' 'IN' 'CR' 'NI' 'HK' 'AD' 'CZ' 'AE' 'MA' 'GR' 'PR' 'RO' 'IL'
 'RU' 'GT' 'GA' 'NO' 'SN' 'MR' 'UA' 'BG' 'PY' 'EE' 'SV' 'ET' 'CM' 'SA'
 'CI' 'QA' 'LU' 'PA' 'BA' 'BO' 'AU' 'BY' 'KE' 'SG' 'HR' 'MD' 'SK' 'TR'
 'AO' 'CU' 'GQ' 'EG' 'ZA' 'DK' 'UY' 'GE' 'TH' 'DZ' 'LB' 'JP' 'NG' 'PK'
 'TN' 'TG' 'KR' 'GH' 'RS' 'VN' 'PH' 'KW' 'NZ' 'MM' 'KH' 'GI' 'SL' 'GN'
 'GW' 'OM' 'CG' 'LV' 'LT' '

tipodom - num unique values: 1 
 [ 1. nan] 


cod_prov - num unique values: 52 
 [29. 13. 50. 45. 24. 20. 10. 17. 49.  8. 37.  9. 22. 31.  5. 40. 27. 25.
 28.  3. 42. 41. 39.  7. 47. 36. 46. 44. 15. 32. 23. 16. 48. 12. 26.  2.
  6. 30. 11. nan  4. 19. 34. 35. 14. 21. 18. 33. 38. 52. 43.  1. 51.] 


nomprov - num unique values: 52 
 ['MALAGA' 'CIUDAD REAL' 'ZARAGOZA' 'TOLEDO' 'LEON' 'GIPUZKOA' 'CACERES'
 'GIRONA' 'ZAMORA' 'BARCELONA' 'SALAMANCA' 'BURGOS' 'HUESCA' 'NAVARRA'
 'AVILA' 'SEGOVIA' 'LUGO' 'LERIDA' 'MADRID' 'ALICANTE' 'SORIA' 'SEVILLA'
 'CANTABRIA' 'BALEARS, ILLES' 'VALLADOLID' 'PONTEVEDRA' 'VALENCIA'
 'TERUEL' 'CORUÑA, A' 'OURENSE' 'JAEN' 'CUENCA' 'BIZKAIA' 'CASTELLON'
 'RIOJA, LA' 'ALBACETE' 'BADAJOZ' 'MURCIA' 'CADIZ' nan 'ALMERIA'
 'GUADALAJARA' 'PALENCIA' 'PALMAS, LAS' 'CORDOBA' 'HUELVA' 'GRANADA'
 'ASTURIAS' 'SANTA CRUZ DE TENERIFE' 'MELILLA' 'TARRAGONA' 'ALAVA' 'CEUTA'] 


ind_actividad_cliente - num unique values: 2 
 [ 1.  0. nan] 


renta - num unique values: 520988 
 

I've found that field ncodpers (which is the personal code) have 956,645 values, so that means the number of customers represented on this data. We have 17th values in "fecha_dato" field; from January 2015 to May 2016. So as there is only 13,647,309 rows, that means some customers won't have all records.

The columns that have low number of values are the ones listed below, in order to understand the data, lets reseach what means each field and if it needs to be cleaned(NBC):

- ind_empleado ==> NBC - Categorical - A active, B ex employed, F filial, N not employee, P pasive.
- sexo ==> NBC - Categorical - H women, V men.
- ind_nuevo ==> NBC - Non categorical, it indicates with '1' if customer got registered in the last 6 months.
- indrel ==> NBC - Categorical - 1 First/Primary, 99 Primary customer during the month but not at the end of the month
- indrel_1mes ==> NBC - Categorical - 1 (First/Primary customer), 2 (co-owner ), P (Potential), 3 (former primary), 4(former co-owner)
- tiprel_1mes ==> NBC - Categorical - Customer relation type at the beginning of the month, A (active), I (inactive), P (former customer), R (Potential)
- indresi ==> NBC - Categorical - Residence index - Same residence country than the bank - S (Yes) or N (No)
- indext ==> NBC - Categorical - Foreigner index - Birth country different than bank's residence country - S (Yes) or N (No)
- conyuemp ==> NBC - Categorical - S (married with an employee) N (not married with an employee)
- indfall ==> NBC - Categorical - S (customer has died) N (customer alive)
- tipodom ==> To delete - Address type - 1 for primary residence
- ind_actividad_cliente ==> NBC - 1, active customer; 0, inactive customer
- segmento ==> NBC - Kind of customer
- ind_ahor_fin_ult1 ==> Target column - Saving account
- ind_aval_fin_ult1 ==> Target column - Guarantees
- ind_cco_fin_ult1 ==> Target column - Current Accounts
- ind_cder_fin_ult1 ==> Target column - Derivada Account
- ind_cno_fin_ult1 ==> Target column - Payroll Account
- ind_ctju_fin_ult1 ==> Target column - Junior Account
- ind_ctma_fin_ult1 ==> Target column - Más particular Account
- ind_ctop_fin_ult1 ==> Target column - particular Account
- ind_ctpp_fin_ult1 ==> Target column - particular Plus Account
- ind_deco_fin_ult1 ==> Target column - Short-term deposits
- ind_deme_fin_ult1 ==> Target column - Medium-term deposits
- ind_dela_fin_ult1 ==> Target column - Long-term deposits
- ind_ecue_fin_ult1 ==> Target column - e-account
- ind_fond_fin_ult1 ==> Target column - Funds
- ind_hip_fin_ult1 ==> Target column - Mortgage
- ind_plan_fin_ult1 ==> Target column - Pensions
- ind_pres_fin_ult1 ==> Target column - Loans
- ind_reca_fin_ult1 ==> Target column - Taxes
- ind_tjcr_fin_ult1 ==> Target column - Credit Card
- ind_valo_fin_ult1 ==> Target column - Securities
- ind_viv_fin_ult1 ==> Target column - Home Account
- ind_nomina_ult1 ==> Target column - Payroll
- ind_nom_pens_ult1 ==> NBC - Target column - Pensions
- ind_recibo_ult1 ==> Target column - Direct Debit

In [13]:
bank_total.dtypes

fecha_dato                object
ncodpers                   int32
ind_empleado              object
pais_residencia           object
sexo                      object
age                       object
fecha_alta                object
ind_nuevo                float32
antiguedad                object
indrel                   float32
ult_fec_cli_1t            object
indrel_1mes               object
tiprel_1mes               object
indresi                   object
indext                    object
conyuemp                  object
canal_entrada             object
indfall                   object
tipodom                  float32
cod_prov                 float32
nomprov                   object
ind_actividad_cliente    float32
renta                    float32
segmento                  object
ind_ahor_fin_ult1           int8
ind_aval_fin_ult1           int8
ind_cco_fin_ult1            int8
ind_cder_fin_ult1           int8
ind_cno_fin_ult1            int8
ind_ctju_fin_ult1           int8
ind_ctma_f

# <font color=blue>TRANSFORMATIONS</font>

#### <font color=blue>We can appreciate that:</font>
<font color=blue>
    
* columns 'fecha_dato' and 'fecha_alta' are classified as object but are dates. I will keep them as objects


* columns 'ind_nuevo' and 'fecha_alta' have exactly the same number of null values (27734). Looking deeper, I've found those nulls are on same rows. As that number of rows represent a 0.20% I'm going to delete it.


* column 'conyuemp' have null values, I'm going to consider them as 'N'.


* column 'age' have spaces on number fields and 27734 fields with NA. Looking further, it get my attention that the number of rows is the same number than 'ind_nuevo' and 'fecha_alta', however I've checked those nulls are not in the same rows. I'm going to convert all column to numbers and drop those NA values since it represent a 0.20% of all data and I'm going to follow "less is more" principle. Additionally, I've found there are some values for underage people and people that have more than 100 years (164 years?? that's not possible). For underage people, may be accounts created by their parents, for "overaged" people it have to be a mistake, so I will move those outliers to the mean of closest one in order to have it more balanced.


* column 'antiguedad' have text cells with NAs with spaces, numbers with spaces and wrong number values(-999999). I'm going to replace all and convert all column to numbers. The wrong number of values and the NA's will be replaced for minimum seniority (0).


* column 'indrel_1mes' have values represented as float, integers... I'm going to standarize it. Additionally, this column haves 149,781 rows with null values, those rows represent a 1,10% of all data. However, inspecting test file I've found there are some nulls as well. So I have to keep those null values.


* column 'tiprel_1mes' have some nan values that represents less that 1,10% of all data. Same as before, I've found on test file some nan values, so I gonna keep them as well.


* column 'segmento' have 161634 rows with null values that represent 1.19% of all data. Same as before, I've found on test file some nan values, so I gonna keep them as well.


* column 'ult_fec_cli_1t' is classified as object, but it have some dates. However, it have lots of nulls so I won't do anything by now.

* For null values in columns 'sexo', 'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'segmento', 'canal_entrada', 'nomprov' - I will create a new category called 'unknown'.


* I will drop 'cod_prov' column since it is exactly the same as 'nomprov'.


* column 'renta' have 2,766,641 missing values that have to be corrected or dropped. Those data represent a 20.31% of all rows. However, taking in account that any row from test file have missing values on 'renta' field and I still have 10,852,934 rows for trainning and testing, I will drop those rows.


* column 'ind_nomina_ult1' and 'ind_nom_pens_ult1' have 177 null values. Having in consideration that any row from test file have missing values on those fields, I will drop those rows.


</font>

In [14]:
a = bank_total[bank_total['ind_nuevo'].isna()].index.tolist()
b = bank_total[bank_total['fecha_alta'].isna()].index.tolist()
c = bank_total[bank_total['age'].isna()].index.tolist()
print(len(a))
print(a == b)
print(a == c)

27734
True
False


In [15]:
bank_total["antiguedad"].isna().sum()

0

ELEMENTS FOR DATA CLEANING

In [16]:
def transformations(df): 
    df['conyuemp'].fillna('N', inplace=True)
    df['age'].replace(' ', '', inplace=True, regex=True)
    #df.loc[df['age']=='NA'] = df['age'].mean()
    df['age'] = pd.to_numeric(df['age'], errors='ignore', downcast='integer')
    df['antiguedad'].replace(' ', '', inplace=True, regex=True)
    df['antiguedad'].replace(' NA', 0, inplace=True)
    df['antiguedad'].replace('NA', 0, inplace=True)
    df['antiguedad'].replace('-999999', 0, inplace=True, regex=True)
    df['antiguedad'] = pd.to_numeric(df['antiguedad'], errors='ignore', downcast='integer')
    df['indrel_1mes'].replace('1.0', '1', inplace=True)
    df['indrel_1mes'].replace('2.0', '2', inplace=True)
    df['indrel_1mes'].replace('3.0', '3', inplace=True)
    df['indrel_1mes'].replace('4.0', '4', inplace=True)
    df['indrel_1mes'].replace(1.0, '1', inplace=True)
    df['indrel_1mes'].replace(2.0, '2', inplace=True)
    df['indrel_1mes'].replace(3.0, '3', inplace=True)
    df['indrel_1mes'].replace(4.0, '4', inplace=True)
    df['indrel_1mes'].replace('1.', '1', inplace=True)
    df['indrel_1mes'].replace('2.', '2', inplace=True)
    df['indrel_1mes'].replace('3.', '3', inplace=True)
    df['indrel_1mes'].replace('4.', '4', inplace=True)
    sp_cols = ['sexo', 'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'segmento', 'canal_entrada', 'nomprov']
    for col in sp_cols: df.loc[df[col].isnull(),col] = "UNKNOWN"
    return "Done"

In [17]:
def nulls_dropper(df):
    df = df.dropna(subset=['ind_nuevo', 'renta', 'ind_nomina_ult1', 'ind_nom_pens_ult1'], inplace=True)
    #df = df.drop(['cod_prov'], axis=1, inplace=True)

    return "Done"
    

In [18]:
nulls_dropper(bank_total)

'Done'

In [19]:
bank_total.drop(['cod_prov'], axis=1, inplace=True)

In [20]:
transformations(bank_total)

'Done'

In [21]:
bank_total.shape

(10852757, 47)

Now, as aforesaid, let's fix the age

In [22]:
age_mean = int(bank_total['age'].mean())

print('avg:',age_mean)

avg: 40


In [23]:
age_values = bank_total['age']
age_values.head()

0    35
1    23
2    23
3    22
5    23
Name: age, dtype: int8

In [24]:
%%time

young_age_mean = int(age_values.loc[(age_values >= 18) & (age_values <= 35)].mean())
print('young:',young_age_mean)

young: 25
CPU times: user 421 ms, sys: 4.56 ms, total: 425 ms
Wall time: 106 ms


In [25]:
%%time

old_age_mean = int(age_values.loc[(age_values >= 50) & (age_values <= 127)].mean())
print('old:',old_age_mean)

old: 63
CPU times: user 363 ms, sys: 3.46 ms, total: 367 ms
Wall time: 91.3 ms


In [26]:
%%time

def age_corrector(df):
    age_mean = 40
    young_age_mean = 25
    old_age_mean = 63
    df.loc[df['age'] =='NA', 'age']  = age_mean
    df.loc[df['age'] ==' NA', 'age']  = age_mean
    df.loc[df['age'] < 18, 'age']  = young_age_mean
    df.loc[df['age'] > 127, 'age'] = old_age_mean
    return 'Done'

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 6.91 µs


In [27]:
age_corrector(bank_total)

/home/alfonso/miniconda3/envs/kaggle/lib/python3.7/site-packages/pandas/core/ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


'Done'

Now, lets check if everything is OK

In [28]:
bank_total.isna().sum()

fecha_dato               0
ncodpers                 0
ind_empleado             0
pais_residencia          0
sexo                     0
age                      0
fecha_alta               0
ind_nuevo                0
antiguedad               0
indrel                   0
ult_fec_cli_1t           0
indrel_1mes              0
tiprel_1mes              0
indresi                  0
indext                   0
conyuemp                 0
canal_entrada            0
indfall                  0
tipodom                  0
nomprov                  0
ind_actividad_cliente    0
renta                    0
segmento                 0
ind_ahor_fin_ult1        0
ind_aval_fin_ult1        0
ind_cco_fin_ult1         0
ind_cder_fin_ult1        0
ind_cno_fin_ult1         0
ind_ctju_fin_ult1        0
ind_ctma_fin_ult1        0
ind_ctop_fin_ult1        0
ind_ctpp_fin_ult1        0
ind_deco_fin_ult1        0
ind_deme_fin_ult1        0
ind_dela_fin_ult1        0
ind_ecue_fin_ult1        0
ind_fond_fin_ult1        0
i

In [29]:
all_columns = ['fecha_dato', 'ncodpers', 'ind_empleado', 'pais_residencia', 'sexo',
       'age', 'fecha_alta', 'ind_nuevo', 'antiguedad', 'indrel',
       'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'indresi', 'indext',
       'conyuemp', 'canal_entrada', 'indfall', 'tipodom',
       'nomprov', 'ind_actividad_cliente', 'renta', 'segmento',
       'ind_ahor_fin_ult1', 'ind_aval_fin_ult1', 'ind_cco_fin_ult1',
       'ind_cder_fin_ult1', 'ind_cno_fin_ult1', 'ind_ctju_fin_ult1',
       'ind_ctma_fin_ult1', 'ind_ctop_fin_ult1', 'ind_ctpp_fin_ult1',
       'ind_deco_fin_ult1', 'ind_deme_fin_ult1', 'ind_dela_fin_ult1',
       'ind_ecue_fin_ult1', 'ind_fond_fin_ult1', 'ind_hip_fin_ult1',
       'ind_plan_fin_ult1', 'ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1']
for col in all_columns:
    print(col, '- num unique values:', bank_total[col].nunique(), '\n', bank_total[col].unique(), '\n\n')

fecha_dato - num unique values: 17 
 ['2015-01-28' '2015-02-28' '2015-03-28' '2015-04-28' '2015-05-28'
 '2015-06-28' '2015-07-28' '2015-08-28' '2015-09-28' '2015-10-28'
 '2015-11-28' '2015-12-28' '2016-01-28' '2016-02-28' '2016-03-28'
 '2016-04-28' '2016-05-28'] 


ncodpers - num unique values: 709411 
 [1375586 1050611 1050612 ... 1432584 1346412 1355691] 


ind_empleado - num unique values: 5 
 ['N' 'A' 'F' 'B' 'S'] 


pais_residencia - num unique values: 25 
 ['ES' 'PA' 'MX' 'IT' 'BE' 'DE' 'BO' 'PY' 'RO' 'GB' 'US' 'CH' 'QA' 'JP'
 'DZ' 'BM' 'LU' 'CA' 'PE' 'PT' 'OM' 'FR' 'NL' 'AD' 'CL'] 


sexo - num unique values: 3 
 ['H' 'V' 'UNKNOWN'] 


age - num unique values: 102 
 [ 35  23  22  65  24  28  25  26  53  27  32  37  31  39  63  33  55  42
  38  50  30  45  44  36  29  60  57  67  47  34  48  46  54  84  83  77
  69  52  59  43  49  41  58  40  51  78  73  66  62  68  82  95  96  56
  61  79  72  19  86  64  20  71  70  74  21  18  75  81  80  76  92  89
  85  91  87  90  94  99  

ind_viv_fin_ult1 - num unique values: 2 
 [0 1] 


ind_nomina_ult1 - num unique values: 2 
 [0. 1.] 


ind_nom_pens_ult1 - num unique values: 2 
 [0. 1.] 


ind_recibo_ult1 - num unique values: 2 
 [0 1] 




### <font color=blue> Cleaning done!! Now I'm going to save the cleaned file. </font>

###### <font color=grey> First I will save it as csv (with column formats in another file) and then I will save it in parquet format.
</font>

In [30]:
file_format = bank_total.dtypes.apply(lambda x: x.name).to_dict()
print(file_format)

{'fecha_dato': 'object', 'ncodpers': 'int32', 'ind_empleado': 'object', 'pais_residencia': 'object', 'sexo': 'object', 'age': 'int8', 'fecha_alta': 'object', 'ind_nuevo': 'float32', 'antiguedad': 'int16', 'indrel': 'float32', 'ult_fec_cli_1t': 'object', 'indrel_1mes': 'object', 'tiprel_1mes': 'object', 'indresi': 'object', 'indext': 'object', 'conyuemp': 'object', 'canal_entrada': 'object', 'indfall': 'object', 'tipodom': 'float32', 'nomprov': 'object', 'ind_actividad_cliente': 'float32', 'renta': 'float32', 'segmento': 'object', 'ind_ahor_fin_ult1': 'int8', 'ind_aval_fin_ult1': 'int8', 'ind_cco_fin_ult1': 'int8', 'ind_cder_fin_ult1': 'int8', 'ind_cno_fin_ult1': 'int8', 'ind_ctju_fin_ult1': 'int8', 'ind_ctma_fin_ult1': 'int8', 'ind_ctop_fin_ult1': 'int8', 'ind_ctpp_fin_ult1': 'int8', 'ind_deco_fin_ult1': 'int8', 'ind_deme_fin_ult1': 'int8', 'ind_dela_fin_ult1': 'int8', 'ind_ecue_fin_ult1': 'int8', 'ind_fond_fin_ult1': 'int8', 'ind_hip_fin_ult1': 'int8', 'ind_plan_fin_ult1': 'int8', 'in

In [31]:
import pickle

with open('./data/fileformat.pickle', 'wb') as handle:
    pickle.dump(file_format, handle, protocol=pickle.HIGHEST_PROTOCOL)

Now, lets save the dataframe:

In [32]:
bank_total.to_csv('./data/clean_train.csv')

In [33]:
import pyarrow as pa

table = pa.Table.from_pandas(bank_total)
type(table)

/home/alfonso/miniconda3/envs/kaggle/lib/python3.7/site-packages/pyarrow/pandas_compat.py:114: FutureWarning: A future version of pandas will default to `skipna=True`. To silence this warning, pass `skipna=True|False` explicitly.
  result = infer_dtype(pandas_collection)


pyarrow.lib.Table

In [34]:
import pyarrow.parquet as pq


pq.write_table(table, './data/clean_train.parquet')

##### After making some tests I decide I have to work with part of dataframe since it is still too heavy, so let's select part of dataframe and save it.

After reviewing other's people kernel, I've found:

* Almost all people has trained without all data provided, just with a percentage they considered, at least Python users.
* Lots of people followed an principle called "When Less is more" established by user BreakfastPirate, that mentions the benefit of using accounts that added a new product on June 2015 as training, since we are only trying to predict which products they would add if they did.
* Lots of people has solved it using Xgboost model.
* Some people has mentioned that recommending the smallest classes wastes recommendations.


In [35]:
selected = ['2015-05-28', '2015-06-28']
bank_sample = bank_total[bank_total['fecha_dato'].isin(selected)]
table_sample = pa.Table.from_pandas(bank_sample)
bank_sample.shape

(1045683, 47)

In [36]:
pq.write_table(table_sample, './data/clean_train_sample.parquet')

Option old

selection = pd.Series(bank_total.ncodpers.unique()).sample(frac=0.1).values
selection

bank_sample = bank_total[bank_total['ncodpers'].isin(selection)]

table_sample = pa.Table.from_pandas(bank_sample)

pq.write_table(table_sample, './data/clean_train_sample.parquet')

### Here I check and transform the test file

In [37]:
test = pd.read_csv('./data/test_ver2.csv')

/home/alfonso/miniconda3/envs/kaggle/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [38]:
test.drop(['cod_prov'], axis=1, inplace=True)

In [39]:
transformations(test)

'Done'

In [40]:
age_corrector(test)

/home/alfonso/miniconda3/envs/kaggle/lib/python3.7/site-packages/pandas/core/ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


'Done'

In [41]:
test.isnull().sum()

fecha_dato               0
ncodpers                 0
ind_empleado             0
pais_residencia          0
sexo                     0
age                      0
fecha_alta               0
ind_nuevo                0
antiguedad               0
indrel                   0
ult_fec_cli_1t           0
indrel_1mes              0
tiprel_1mes              0
indresi                  0
indext                   0
conyuemp                 0
canal_entrada            0
indfall                  0
tipodom                  0
nomprov                  0
ind_actividad_cliente    0
renta                    0
segmento                 0
dtype: int64

In [42]:
test.shape

(929615, 23)

In [43]:
test.dtypes

fecha_dato               object
ncodpers                  int64
ind_empleado             object
pais_residencia          object
sexo                     object
age                       int16
fecha_alta               object
ind_nuevo                 int64
antiguedad                int32
indrel                    int64
ult_fec_cli_1t           object
indrel_1mes              object
tiprel_1mes              object
indresi                  object
indext                   object
conyuemp                 object
canal_entrada            object
indfall                  object
tipodom                   int64
nomprov                  object
ind_actividad_cliente     int64
renta                    object
segmento                 object
dtype: object

In [44]:
test['indrel_1mes'].unique()

array(['1', '3', 'UNKNOWN'], dtype=object)

In [45]:
bank_total['indrel_1mes'].unique()

array(['1', '3', 'UNKNOWN', '2', 'P', '4'], dtype=object)

In [46]:
test_clean = pa.Table.from_pandas(test)

/home/alfonso/miniconda3/envs/kaggle/lib/python3.7/site-packages/pyarrow/pandas_compat.py:114: FutureWarning: A future version of pandas will default to `skipna=True`. To silence this warning, pass `skipna=True|False` explicitly.
  result = infer_dtype(pandas_collection)


In [47]:
pq.write_table(test_clean, './data/clean_test.parquet')

In [49]:
test.to_csv('./data/clean_test.csv')